plugins/vpp/ipsecplugin/vppcalls/dump_vppcalls.go

```diff
@@ -0,0 +1,385 @@
+// Copyright (c) 2018 Cisco and/or its affiliates.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package vppcalls
+
+import (
+	"bytes"
+	"net"
+	"strconv"
+	"time"
+
+	ipsecapi "github.com/ligato/vpp-agent/plugins/vpp/binapi/ipsec"
+	"github.com/ligato/vpp-agent/plugins/vpp/model/interfaces"
+	"github.com/ligato/vpp-agent/plugins/vpp/model/ipsec"
+)
+
+// IPSecSaDetails holds security association with VPP metadata
+type IPSecSaDetails struct {
+	Sa   *ipsec.SecurityAssociations_SA
+	Meta *IPSecSaMeta
+}
+
+// IPSecSaMeta contains all VPP-specific metadata
+type IPSecSaMeta struct {
+	SaID           uint32
+	Interface      string
+	IfIdx          uint32
+	CryptoKeyLen   uint8
+	IntegKeyLen    uint8
+	Salt           uint32
+	SeqOutbound    uint64
+	LastSeqInbound uint64
+	ReplayWindow   uint64
+	TotalDataSize  uint64
+}
+
+func (handler *ipSecVppHandler) DumpIPSecSA() (saList []*IPSecSaDetails, err error) {
+	return handler.DumpIPSecSAWithIndex(^uint32(0)) // Get everything
+}
+
+func (handler *ipSecVppHandler) DumpIPSecSAWithIndex(saID uint32) (saList []*IPSecSaDetails, err error) {
+	defer func(t time.Time) {
+		handler.stopwatch.TimeLog(ipsecapi.IpsecSaDump{}).LogTimeEntry(time.Since(t))
+	}(time.Now())
+
+	saDetails, err := handler.dumpSecurityAssociations(saID)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, saData := range saDetails {
+		// Skip tunnel interfaces
+		if saData.SwIfIndex != ^uint32(0) {
+			continue
+		}
+
+		// Addresses
+		var tunnelSrcAddrStr, tunnelDstAddrStr string
+		if uintToBool(saData.IsTunnelIP6) {
+			var tunnelSrcAddr, tunnelDstAddr net.IP = saData.TunnelSrcAddr, saData.TunnelDstAddr
+			tunnelSrcAddrStr, tunnelDstAddrStr = tunnelSrcAddr.String(), tunnelDstAddr.String()
+		} else {
+			var tunnelSrcAddr, tunnelDstAddr net.IP = saData.TunnelSrcAddr[:4], saData.TunnelDstAddr[:4]
+			tunnelSrcAddrStr, tunnelDstAddrStr = tunnelSrcAddr.String(), tunnelDstAddr.String()
+		}
+
+		sa := &ipsec.SecurityAssociations_SA{
+			Spi:            saData.Spi,
+			Protocol:       getSaProto(saData.Protocol),
+			CryptoAlg:      getCryptoAlg(saData.CryptoAlg),
+			CryptoKey:      string(bytes.SplitN(saData.CryptoKey, []byte{0x00}, 2)[0]),
```

Comment:
I believe that this should not be split by zero byte and should use `CryptoKeyLen` field to use specific number of bytes.

Just like it is done for configuring: https://github.com/ligato/vpp-agent/blob/ea8e83b4c762a466bf1dbde1638a7cb34fe8e229/plugins/vpp/ipsecplugin/vppcalls/ipsec_vppcalls.go#L85